File: app/src/main/kotlin/com/spacex/ui/rockets/RocketsPresenter.java
package com.spacex.ui.rockets;

import androidx.lifecycle.Lifecycle.Event;
import androidx.lifecycle.LifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.OnLifecycleEvent;
import com.spacex.ui.rockets.RocketsContract.Presenter;
import com.spacex.ui.rockets.RocketsContract.View;

public class RocketsPresenter implements Presenter, LifecycleObserver {

  private final RocketsContract.View view;

  RocketsPresenter(final View view) {
    this.view = view;
    if (view instanceof LifecycleOwner) {
      ((LifecycleOwner) view).getLifecycle().addObserver(this);
    }
  }

  @Override
  @OnLifecycleEvent(Event.ON_CREATE)
  public void onCreate() {
    view.setUpViews();
  }

  @Override
  public void onUpNavigation() {
    view.navigateUp();
  }
}

File: app/src/main/kotlin/com/spacex/ui/rockets/RocketsPresenter.kt
package com.spacex.ui.rockets

import androidx.lifecycle.Lifecycle.Event
import androidx.lifecycle.LifecycleObserver
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.OnLifecycleEvent
import com.spacex.ui.rockets.RocketsContract.Presenter
import com.spacex.ui.rockets.RocketsContract.View

internal class RocketsPresenter(
        private val view: View,
        private val selectedRocketId: String,
        private val rocketIds: ArrayList<String>) : Presenter, LifecycleObserver {

    init {
        if (view is LifecycleOwner) {
            (view as LifecycleOwner).lifecycle.addObserver(this)
        }
    }

    @OnLifecycleEvent(Event.ON_CREATE)
    override fun onCreate() {
        view.setUpViews(selectedRocketId, rocketIds)
    }

    override fun onUpNavigation() {
        view.navigateUp()
    }
}
